### DISCUSSION TITLE
Screen reader accessibility for Ghostty

### DISCUSSION THREAD

**Discussion by mikolysz** [ghostty-org/ghostty#2351] [Feature Requests, Ideas]
This is a general issue for tracking the state of Ghostty's screen-reader support. I will try to explain what needs to be done as best as I can, along with some pointers on what to do and what pitfalls to avoid. We might want to create specific issues for the items described here at some point.

## Current status

At this time, accessibility support in Ghostty is basically nonexistent. This is mostly due to the fact that Ghostty uses GPU rendering, which prevents screen readers from extracting any information from the app's UI tree as they otherwise would. However, this issue can be worked around in multiple ways (see below). Here's a list of things that need to be done, roughly in this order:

- [ ] Implement direct screen reader output
- [ ] Expose the current terminal state to assistive technologies
- [ ] Implement AT support for cursor navigation, as used in text editors and other such software
- [ ] Allow for temporarily silencing speech output, such as for long-running and spammy commands like `brew install`.
- [ ] Implement custom punctuation handling, might not be needed depending on how we do speech in the first place

This list is by no means exhaustive, it's merely a baseline of features which are absolutely essential for Ghostty to be AT-friendly. If we're done with these, we can think about nice-to-haves and things that make our lives a lot more pleasant but can be lived without.

Below are a few notes on how those features can be implemented and the tradeoffs involved.

## Speech output

Whenever we type a command, we would expect the results of said command to be announced by our screen reader. The obvious way to do this is to communicate with the screen readers directly, most of them expose an API for this. Windows has libraries that allow us to do all of this with a unified API, instead of having to implement five or so separate APIs. However, most of them would require us to ship (redistributable) DLLs that come from the screen reader vendors, some of which we don't have source code for as far as I believe. On the Mac side, communications with Voice Over (basically the only screen reader that exists) can be achieved through AppleScript. I don't know if there's a better way, perhaps we could somehow invoke the actions that apps expose to AppleScript directly without going through the whole scripting layer? I don't really know what the situation is on the Linux side, as I barely ever use desktop Linux, we'd have to ask somebody more knowledgeable about how Linux GUI accessibility works.

However, this method has a huge drawback, particularly on Mac OS. On most other platforms, screen readers have some kind of speech queue. When requesting a screen reader to speak a particular string, we can usually choose whether the string should be spoken immediately, interrupting all other speech, or pushed to the end of the queue. Voice Over on the Mac unfortunately doesn't offer this functionality, all new speech interrupts whatever is currently being spoken. This feature is very important for a terminal app, as we're often dealing with apps that produce large quantities of output, often generating new text several times a second. In such a situation, doing things the Voice Over way would prevent us from understanding anything, as each message would instantly be interrupted by the beginning of the next one. We couldn't even manually review terminal output, as the same issue would occur. The native terminal app has this problem, so the vast majority of people don't use it with VO. Maybe there would be a way to hack around this by doing audio trickery and figuring out if voice over is still speaking or not, pushing new messages when it's done, but I'm not very hopeful there and I wouldn't recommend that strategy. Another issue is that Voice Over's Apple Script support needs to be explicitly enabled in Voice Over utility, and this setting is very flaky across system updates, it often shows as enabled while it actually isn't and has to be re-enabled manually.

One way to mitigate this problem is to interact with the system speech APIs directly, completely bypassing the screen reader. Those APIs do support queuing, and in the worst case scenario, we can always request raw audio buffers and play them in whatever way we see fit. This is what I'd recommend to do on Mac, there's really no other solution here. However, this approach solves one big problem but introduces two smaller ones, voice handling and punctuation schemas.

If we're using the system speech API, we have to allow the user to adjust their voice parameters, such as the actual voice to use, the speech rate, pitch, volume, language etc. We can implement configuration for this, forcing the user to set the parameters themselves, but a more interesting approach would be to extract that information directly from the screen reader, making the user completely unaware that we're actually generating our speech without its help. This would be difficult on Mac OS 13 and would require us to parse Voice Over's plist files, but Sonoma will expose an API for this. Perhaps we can hold off on this for now and just use the default system voice, and then switch to that API once Sonoma is stable. To be clear, we could probably find a library that does all of this for us, including cross-platform support, I'd have to check on what's good these days.

Another issue is punctuation handling. Most people who use terminals interact with code, so they want punctuation characters to be announced, something that the system speech API won't do by default. We'd basically have to implement punctuation handling ourselves and replace characters like ":" and "?" with words like "colon" and "question". The crude way to do this is with a static mapping, however, that doesn't let the user change the character descriptions, such as when their text is often in another language and they use a non-english voice, or when they care about efficiency and want "is" instead of "equals" or "score" instead of "underscore". Again, we can either implement configuration for this or parse the sqlite files where Voice Over's punctuation schemas are kept.

## UI state

Linear speech output isn't everything, we often also need to review the output history in detail. Complicated pieces of code can't be understood all at once, there has to be a way for us to navigate line-by-line, word-by-word etc, hearing the content we move past. Once again, there are two ways to go about this:

### Exposing the UI state to platform accessibility APIS

As I said above, Ghostty doesn't currently expose anything to the platform APIs because of the GPU rendering. Those APIs expect a tree of native widgets which are accessible out of the box, but which Ghostty doesn't utilize. However, it's not the first project with that problem, the same hurdle has been faced by third-party GUI frameworks like QT or Java's Swing, as well as browser engines. For that reason, platforms let you expose accessibility objects even for things that aren't native widgets, something that those frameworks and browser engines must do. We could expose the terminal contents this way. There might be a performance impact, but there are APIs to detect if there's  active assistive technology software consuming this info, so we wouldn't need to do this work for fully-abled users. Until recently, implementing something like that required us to do the work separately for each platform, dealing with COM interfaces on Windows and underdocumented ATSPI on Linux. However, there's now a [library that abstracts those platform APIs behind a common abstraction](https://github.com/AccessKit/accesskit). The library is quite new, and I'm not sure if it implements everything we need, but it's [already used by one somewhat popular Rust UI framework](https://news.ycombinator.com/item?id=33859697). The library exposes a C API but is written in Rust, not sure if this is acceptable to us and what impact it would have on our build process.

This approach would be preferable, as it lets users use the screen reader commands they already know. In addition, it also exposes the terminal contents to non-screen-reader assistive technologies, scripting and automation apps such as Hammerspoon on Mac and AutoID on Windows and allows reviewing output using refreshable braille displays, something that would be appreciated by users who are also hard of hearing or who just prefer to do things that way. Screen readers need to get the terminal contents from the native APIs if we want braille displays to work.

This way of doing things would also let us simplify how new output is delivered to the screen reader (at least outside of Mac OS). Most platforms provide a way to mark a text widget as a "live region", indicating to the screen reader that any new content that appears should be spoken automatically. Voice Over on Mac still has the "no speech queue" problem, so system speech seems to be the only way here. Outside of Mac OS, however, this approach might even let us dispose of a `speak_string` function completely.

## Implementing a micro screen-reader

Instead of exposing content to accessibility APIs, we can fully lean on the "speak output" function explained in the previous section. We can implement commands like "next line", "previous character" etc manually, give them custom keyboard shortcuts, figure out what needs to be said (probably the line/word/character we just passed) and speak the constructed message ourselves. This is how VS Code's terminal accessibility is implemented, although in a really terrible way, same goes for Warp, which is even worse. However, [many](https://github.com/tspivey/tdsr) [terminal](https://github.com/chrys87/fenrir) [screen](https://brltty.app/) [readers](http://www.linux-speakup.org/) do this successfully. This approach, however, doesn't provide Braille support, doesn't work well with other AT software and is generally clunkier and more hacky. I don't know which way would require more work, exposing terminal contents seems like a dauting task, but the micro screen-reader would require us to implement scrolling, selection, copying, announcing line indentation, skipping repeated punctuation and other such features manually, while existing screen readers just give them to us for free. 

## Cursor movement

The next problem we need to tackle is software that relies on cursor navigation, such as text editors, Ncurses menus or even readline prompts. Whenever the cursor moves, we need to announce the right entity, such as the line we moved to or the character we moved past. This isn't as easy as it seems. It's not always obvious when cursor movement is due to a user action and when it's due to the program doing something. Some software resets the cursor after being done with displaying the output, and we should take care not to suppress any speech in that case. We can rely on arrow keys to detect when movement is user initiated, but then software like Vim (with navigation via hjkl) or even the common Readline shortcuts are problematic. Most screen readers (at least NVDA and TDSR) rely on arrow keys and just read whatever entity the cursor is at after a timeout, regardless of whether it moved or not, but this is not great when SSH servers are involved. When the timeout is too low and the server is too slow, the wrong thing is read, but when the timeout is too high, navigating through a file feels extremely frustrating. It would be great to innovate in this area and find a better solution.

As I said above, there's more to be done, the area of terminal accessibility is woefully under explored, and due to the prevalence of GUI platforms, some ideas common to 80s DOS screen readers are pretty much forgotten now. However, if we can get the things described above implemented, we'd be competitive with most of the Windows solutions, better than what we have on the Mac, and still a good bit behind what you can get in a true non-GUI, kernel-based Linux TTY.

**Comment by mitchellh** [comment]
I haven't read the full issue text yet, but I really appreciate the writeup @mikolysz and look forward to giving this a shot. I agree that as specific tasks are discovered, we can break them out. I'd prefer to not break them out until someone is actively working on that task, since this encompasses the pending work nicely, I think.

I know macOS has a LOT of accessibility APIs so I think targeting macOS as the primary accessibility target makes sense. Linux can be indefinitely delayed until its clear what improvements we can make.

**Reply by grishy** [reply]
I made a small breakdown of the ticket into three APIs that are missing or improperly implemented.
https://github.com/ghostty-org/ghostty/discussions/9930

**Comment by alex19EP** [comment]
Quick update from the Linux side. I've been working on a GTK apprt accessibility PoC on my fork (`alex19EP/ghostty` branch [`gtk-accessibility-poc`](https://github.com/alex19EP/ghostty/tree/gtk-accessibility-poc)). As someone who uses Orca and a braille display daily, I needed this myself.

**Approach.** GTK4's native `GtkAccessibleText` interface, not AccessKit. Reasoning: zig-gobject already exposes the GTK a11y vtables, so there's no Rust dependency and no C API shim to maintain — it stays in Zig inside `src/apprt/gtk/class/surface.zig`. This path also feeds the existing AT-SPI bridge that GTK apps already use, so Orca picks us up with no special casing.

**What works today, driving with Orca + a braille display on GNOME:**
- Terminal contents are exposed and read correctly.
- Flat review works end-to-end — KP_7/8/9 line nav, KP_4/5/6 word/char nav.
- Typing echo and command output announce live. The per-frame diff is gated to emit targeted `insert`/`remove` events for just the changed range, not whole-viewport updates.
- Braille display reflects terminal contents.
- Per-character SGR attributes are emitted as AT-SPI attribute runs.
- Selection — `get_selection` / `set_selection` are wired to the terminal's selection state.
- Focus changes between surfaces are announced.
- Inline OSC 8 link announcement via `GtkAccessibleHypertext` (GTK 4.22+, dlsym'd so older GTK still builds). Configured regex links too.
- Cursor position on leading indentation.

**Known tradeoffs.**
- Link activation via AT is deliberately not wired; reasoning in the branch notes.
- A couple of upstream Orca bugs surfaced along the way and are being filed separately.

Sharing this for visibility rather than as a PR proposal — I'd like feedback on whether the approach is the right direction before investing further, and specifically whether a unification with #11196 (`Screen.createAccessibilityContext`) would be welcome: the GTK side could be rebased onto that once it lands, giving macOS and Linux a shared snapshot path.

---

*Disclosure: this PoC was built with extensive help from Claude Code. I'm blind, and this work is driven by my own daily need for an accessible terminal on Linux — I directed the design, tested extensively with Orca and a braille display, filed the upstream Orca bugs I hit, and reviewed the code, but I'm not a Zig developer, so the implementation itself was largely AI-generated under my direction. I'm sharing as a discussion contribution, not yet as a PR proposal, specifically because of Ghostty's "must fully understand your code" rule — I'd want to either deepen my own understanding of the Zig implementation or pair with a Zig contributor before proposing a merge.*

**Comment by alex19EP** [comment]
Hi @mitchellh 

I've been daily-driving a GTK accessibility branch for a couple of weeks and the rough edges are gone. Before I open a PR (or several), I'd like your read on a handful of scope and policy questions so the review doesn't surface them late.

### What it does today

On the GTK apprt, the surface widget implements `GtkAccessibleText` (and, on GTK 4.22+, `GtkAccessibleHypertext`). With Orca running:

- Terminal contents are read; flat review (line/word nav) works.
- Typing echo and command output announce correctly.
- Focus changes between surfaces are announced.
- OSC 8 hyperlinks and configured regex links announce inline
- Selection is exposed read AND write (the GTK 4.22 `set_selection` vfunc lands a real terminal selection via `Screen.select`).
- Caret set by AT clients translates to a synthetic mouse click on the cell.
- Per-character text attributes (fg/bg color, weight, italic, underline, plus `ghostty-*` keys) and default attributes (family, live point size) are reported.
- Braille works (tested throughout development).

Three Orca upstream fixes the work depends on are merged.

### Files touched

- `src/apprt/gtk/class/surface.zig` — bulk of the work: vfuncs, viewport snapshot cache, per-frame diff that emits targeted `.insert`/`.remove`.
- `src/apprt/gtk/class/window.zig` — `accessible-role = window` as a construct-time property.
- `src/apprt/gtk/class/application.zig` — `glib.setPrgname/setApplicationName` before `adw.init()`; quick-terminal role.
- `src/apprt/gtk/ui/1.2/surface.blp` — focus moved from inner `GLArea` to the template root so the focused AT-SPI object carries the role and Text/Hypertext interfaces.
- `src/apprt/gtk/a11y_hypertext.zig` — runtime `std.DynLib` shims for GTK 4.22's `GtkAccessibleHypertext` (zig-gobject 0.3.0 doesn't expose it yet).
- `src/build/SharedDeps.zig` — adds the `graphene` zig-gobject bindings entry.

### What I'd like your read on

**Scope / shape of the PR(s)**
- Single PR, or split into stages?
- Coordination with @MrMage's #11196 (`Screen.createAccessibilityContext`) and #10992 — do you want me to wait for #11196 so the GTK side reuses the same core helper, rebase onto it later, or land independently and converge after?
- Do you want UTF-8 - codepoint helpers (and the snapshot/diff itself) in core (`Surface.zig` / `terminal/Screen.zig`) so macOS can use them, or kept in the GTK apprt for now?

**Versions / dependencies**
- GTK minimum: keep the runtime-optional dlsym for the 4.22 Hypertext path, or hard-bump min GTK so the bindings can be regenerated and `a11y_hypertext.zig` deleted?

**Behavior calls that are yours**
- Blueprint focus moved from inner `GLArea` to template root. Necessary because the AT-SPI focused-object must carry our role + Text interface, but it changes which widget keyboard/IME events route through. I can't test whether this causes a regression without an A11Y client, because I can't use a computer without Orca.
- `axGetFirstAccessibleChild` returns `null` so the surface looks like a single text-bearing leaf in the a11y tree (no `GLArea`/template descendants leak into Orca). Comfortable with that as precedent?
- Content-change emission: per-frame diff in `glareaRender`, gated on `ax_active` (set on first AT query). I deliberately did not adopt #10992's push-style `content_changed` apprt action because of your CPU-spin concern. When AT is present, the diff is one common-prefix/common-suffix scan over the viewport snapshot; happy to try bench numbers if you want them in the PR.
- `axSetCaretPosition` is implemented by synthesizing a mouse click at the target cell. It's the only path that gives correct selection/scroll behavior given the current core API. Acceptable, or would you prefer a dedicated apprt action?

**Code organization**
- The a11y code in `surface.zig` is sizable. i want to  Split into `class/surface_a11y.zig` (or `class/a11y/`)

**Testing**
- Any meaningful CI coverage you'd want? At minimum, a smoke test that the interface registers and `axGetContents` returns sane bytes for a fixture surface. Beyond that, the gotchas are mostly enforced by structure — happy to add asserts where they help.

**Reply by jcollie** [reply]
On CI: as much as possible! The GTK apprt is particularly deficient in tests (not that GTK makes this easy).

I'd hold off on implementing anything from GTK 4.22 until we can update `zig-gobject` (which largely depends on NixOS getting Gnome 50 merged).

I'd break this up into smaller PRs that build on each other as much as possible. That makes it WAY easier to review.
